### ISSUE TITLE
Multiple GPU Mask R-CNN and Faster R-CNN error with torch.nn.DataParallel [Solution Found]

### ISSUE THREAD

**Issue by GerardMaggiolino** [pytorch/vision#1659]
**The current torchvision Mask R-CNN model does not work with multiple GPUs through torch.nn.DataParallel.**

The bug, along with some offered solutions that are ineffective, [are found in this thread](https://discuss.pytorch.org/t/mask-rcnn-nn-dataparallel-possible/53265/5). As this is an issue with the RPN implementation, Faster R-CNN must also be impacted (although I did not test this explicitly). 

There is one warning and one error which must be fixed. The error is a result of AnchorGenerator returning tensors on the wrong device. **This is fixed by adding the following line to /site/package/path/torchvision/models/detection/rpn.py on line 409:** 
```python
anchors = [a.to(pred_bbox_deltas.device) for a in anchors]
```
This should be placed right above this call, which is already in the code:  
```        
proposals = self.box_coder.decode(pred_bbox_deltas.detach(), anchors)
````

This forces all tensors to be on the same device of the bounding box deltas. 

After this, a warning is printed: 
torch/nn/parallel/_functions.py:61: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
This seems to be harmless. 

It should be noted that after doing this the values of the loss dict returned from Mask R-CNN will be a vector with each index corresponding to the device which ran that part of the batch. To accumulate these losses, I'd recommend doing the following: 
```
loss_dict = model(tensors, labels)
# v is now a vector of the loss for that part of the mini-batch
loss_dict = {k: v.mean() for k, v in loss_dict.items()}
```
There's a deeper error with the AnchorGenerator which I didn't check out, but this solution works for now.

**Comment by fmassa** [comment]
You'll be facing other problems down the road by trying to use `nn.DataParallel` with Faster R-CNN. The first being that the input is a list of tensors, and the list will not be splitted correctly.
As such, and because `DataParallel` is significantly slower than `DistributedDataParallel`, we don't have plans to modify the current implementation to support `DataParallel` (the implementation would need to be more complicated, we wouldn't be able to use lists, etc)

I would strongly recommend using `nn.DistributedDataParallel` for training models with multiple GPUs. All models in torchvision support it natively, and training will be much faster as well.

You can use the [reference scripts](https://github.com/pytorch/vision/tree/master/references/detection) to have an idea on how to use the detection models with `DistributedDataParallel`.
